File: core/scanner/scanner.go
package scanner

import (
	"context"
	"sync"
	"time"

	"github.com/agent-toolkit/core"
)

// Scanner 实现了core.Scanner接口
type Scanner struct {
	agents map[string]core.Agent
	mu     sync.RWMutex
}

// NewScanner 创建一个新的Scanner实例
func NewScanner() *Scanner {
	return &Scanner{
		agents: make(map[string]core.Agent),
	}
}

// Scan 扫描所有本地Agent
func (s *Scanner) Scan() ([]core.Agent, error) {
	s.mu.RLock()
	defer s.mu.RUnlock()

	agents := make([]core.Agent, 0, len(s.agents))
	for _, agent := range s.agents {
		agents = append(agents, agent)
	}
	return agents, nil
}

// Watch 监控Agent变化
func (s *Scanner) Watch(ctx context.Context) (<-chan core.AgentEvent, error) {
	ch := make(chan core.AgentEvent)

	go func() {
		defer close(ch)
		ticker := time.NewTicker(1 * time.Second)
		defer ticker.Stop()

		for {
			select {
			case <-ctx.Done():
				return
			case <-ticker.C:
				// 模拟扫描变化
				s.mu.RLock()
				agentCount := len(s.agents)
				s.mu.RUnlock()

				if agentCount == 0 {
					// 模拟添加一个Agent
					agent := core.Agent{
						ID:          "mock-agent-1",
						Name:        "Mock Agent",
						Type:        "mock",
						Status:      "running",
						PID:         12345,
						CommandLine: "mock-agent --config config.yaml",
						StartedAt:   time.Now(),
						Metadata:    map[string]string{"version": "1.0.0"},
					}

					s.mu.Lock()
					s.agents[agent.ID] = agent
					s.mu.Unlock()

					select {
					case ch <- core.AgentEvent{
						Type:  "add",
						Agent: agent,
					}:
					case <-ctx.Done():
						return
					}
				}
			}
		}
	}()

	return ch, nil
}

// AddAgent 添加一个Agent（用于测试和模拟）
func (s *Scanner) AddAgent(agent core.Agent) {
	s.mu.Lock()
	defer s.mu.Unlock()
	s.agents[agent.ID] = agent
}

// RemoveAgent 移除一个Agent
func (s *Scanner) RemoveAgent(id string) {
	s.mu.Lock()
	defer s.mu.Unlock()
	delete(s.agents, id)
}

// GetAgent 获取一个Agent
func (s *Scanner) GetAgent(id string) (core.Agent, bool) {
	s.mu.RLock()
	defer s.mu.RUnlock()
	agent, ok := s.agents[id]
	return agent, ok
}

File: core/scanner/scanner.py
"""Scanner模块的Python实现"""

import asyncio
import time
from typing import List, Dict, Any, AsyncGenerator, Optional


class Scanner:
    """Scanner接口的Python实现"""
    
    def __init__(self):
        self.agents: Dict[str, Dict[str, Any]] = {}
    
    def scan(self) -> List[Dict[str, Any]]:
        """扫描所有本地Agent"""
        return list(self.agents.values())
    
    async def watch(self, timeout: float = 1.0) -> AsyncGenerator[Dict[str, Any], None]:
        """监控Agent变化"""
        # 只运行一次，模拟添加一个agent后退出
        await asyncio.sleep(timeout)
        
        # 模拟扫描变化
        if len(self.agents) == 0:
            # 模拟添加一个Agent
            agent = {
                "id": "mock-agent-1",
                "name": "Mock Agent",
                "type": "mock",
                "status": "running",
                "pid": 12345,
                "command_line": "mock-agent --config config.yaml",
                "started_at": time.time(),
                "metadata": {"version": "1.0.0"},
            }
            
            self.agents[agent["id"]] = agent
            
            yield {
                "type": "add",
                "agent": agent,
            }
    
    def add_agent(self, agent: Dict[str, Any]) -> None:
        """添加一个Agent"""
        self.agents[agent["id"]] = agent
    
    def remove_agent(self, agent_id: str) -> None:
        """移除一个Agent"""
        if agent_id in self.agents:
            del self.agents[agent_id]
    
    def get_agent(self, agent_id: str) -> Optional[Dict[str, Any]]:
        """获取一个Agent"""
        return self.agents.get(agent_id)


class MockScanner:
    """模拟的Scanner实现"""
    
    def __init__(self, agents: List[Dict[str, Any]] = None):
        self.agents = agents or []
    
    def scan(self) -> List[Dict[str, Any]]:
        """返回模拟的Agent列表"""
        return self.agents
    
    async def watch(self, timeout: float = 1.0) -> AsyncGenerator[Dict[str, Any], None]:
        """模拟监控"""
        while True:
            await asyncio.sleep(timeout)
            # 模拟事件
            for agent in self.agents:
                yield {
                    "type": "update",
                    "agent": agent,
                }